File: packages/owl_haptics/android/src/main/kotlin/com/owlllwo/owl_haptics/OwlHapticPlaybackFence.kt
package com.owlllwo.owl_haptics

internal data class OwlHapticPlaybackToken(
    val owner: String,
    val id: Long,
    val generation: Long,
)

internal class OwlHapticPlaybackFence {
    var current: OwlHapticPlaybackToken? = null
        private set

    private var generation = 0L

    fun replace(owner: String, id: Long): OwlHapticPlaybackToken {
        generation = if (generation == Long.MAX_VALUE) 1 else generation + 1
        return OwlHapticPlaybackToken(owner, id, generation).also { current = it }
    }

    fun matches(token: OwlHapticPlaybackToken): Boolean = current == token

    fun matching(owner: String, id: Long): OwlHapticPlaybackToken? =
        current?.takeIf { it.owner == owner && it.id == id }

    fun matching(owner: String): OwlHapticPlaybackToken? =
        current?.takeIf { it.owner == owner }

    fun clear(token: OwlHapticPlaybackToken): Boolean {
        if (current != token) return false
        current = null
        return true
    }
}

File: packages/owl_haptics/ios/owl_haptics/Sources/owl_haptics/OwlHapticPlaybackFence.swift
import Foundation

struct OwlHapticPlaybackToken: Equatable {
  let owner: UUID
  let id: Int64
  let generation: UInt64
}

final class OwlHapticPlaybackFence {
  private(set) var current: OwlHapticPlaybackToken?
  private var generation: UInt64 = 0

  @discardableResult
  func replace(owner: UUID, id: Int64) -> OwlHapticPlaybackToken {
    generation = generation == UInt64.max ? 1 : generation + 1
    let token = OwlHapticPlaybackToken(owner: owner, id: id, generation: generation)
    current = token
    return token
  }

  func matches(_ token: OwlHapticPlaybackToken) -> Bool {
    current == token
  }

  func matching(owner: UUID, id: Int64) -> OwlHapticPlaybackToken? {
    guard let current, current.owner == owner, current.id == id else { return nil }
    return current
  }

  func matching(owner: UUID) -> OwlHapticPlaybackToken? {
    guard let current, current.owner == owner else { return nil }
    return current
  }

  @discardableResult
  func clear(_ token: OwlHapticPlaybackToken) -> Bool {
    guard current == token else { return false }
    current = nil
    return true
  }
}
